external-crates/move/crates/move-binary-format/src/compatibility.rs

```diff
@@ -109,77 +302,88 @@ impl Compatibility {
+            ) {
+                context.struct_ability_mismatch(old_struct, new_struct);
+            }
+
+            if !datatype_type_parameters_compatible(
+                // type param mismatch
                 self.disallow_change_datatype_type_params,
                 &old_struct.type_parameters,
                 &new_struct.type_parameters,
             ) {
-                datatype_and_function_linking = false;
+                context.struct_type_param_mismatch(old_struct, new_struct);
             }
+            // create a diff of the fields?
             if new_struct.fields != old_struct.fields {
                 // Fields changed. Code in this module will fail at runtime if it tries to
                 // read a previously published struct value
                 // TODO: this is a stricter definition than required. We could in principle
                 // choose that changing the name (but not position or type) of a field is
                 // compatible. The VM does not care about the name of a field
                 // (it's purely informational), but clients presumably do.
-                datatype_layout = false
+
+                // field layout mismatch
+                context.struct_field_mismatch(old_struct, new_struct);
             }
         }
 
+        // enums relatively similar but has a variant check
+        // variants have fields which need checking
+        // store
         for (name, old_enum) in &old_module.enums {
             let Some(new_enum) = new_module.enums.get(name) else {
                 // Enum not present in new. Existing modules that depend on this enum will fail to link with the new version of the module.
                 // Also, enum layout cannot be guaranteed transitively, because after
                 // removing the enum, it could be re-added later with a different layout.
-                datatype_and_function_linking = false;
-                datatype_layout = false;
-                break;
+
+                context.enum_missing(old_enum);
+                continue;
             };
 
             if !datatype_abilities_compatible(
                 self.disallowed_new_abilities,
                 old_enum.abilities,
                 new_enum.abilities,
-            ) || !datatype_type_parameters_compatible(
+            ) {
+                context.enum_ability_mismatch(old_enum, new_enum);
+            }
+
+            if !datatype_type_parameters_compatible(
                 self.disallow_change_datatype_type_params,
                 &old_enum.type_parameters,
                 &new_enum.type_parameters,
             ) {
-                datatype_and_function_linking = false;
+                context.enum_type_param_mismatch(old_enum, new_enum);
             }
 
             if new_enum.variants.len() > old_enum.variants.len() {
-                no_new_variants = false;
-            }
-
-            if new_enum.variants.len() < old_enum.variants.len() {
```

Comment:
(Noting for other reviewers that we lost this because the loop below already performs effectively the same check, by a pigeonhole argument).